<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Комментарий-ревьювера" data-toc-modified-id="Комментарий-ревьювера-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Комментарий ревьювера</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Ознакомление-с-таблицами" data-toc-modified-id="Ознакомление-с-таблицами-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Ознакомление с таблицами</a></span></li><li><span><a href="#Задания" data-toc-modified-id="Задания-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задания</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Финальный-комментарий" data-toc-modified-id="Финальный-комментарий-7"><span class="toc-item-num">7&nbsp;&nbsp;</span><span style="color: green">Финальный комментарий<span></span></span></a></span></li></ul></div>

# Выпускной проект по SQL   


**Задача — проанализировать базу данных для крупного сервиса для чтения книг по подписке.**


## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

![Схема данных](https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?table=block&id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=1640&userId=&cache=v2)

## Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Ознакомление с таблицами

In [3]:
books = ''' SELECT * 
              FROM books '''
pd.io.sql.read_sql(books, con = engine).head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
authors = ''' SELECT * 
                FROM authors '''
pd.io.sql.read_sql(authors, con = engine).head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
publishers = ''' SELECT * 
                   FROM publishers '''
pd.io.sql.read_sql(publishers, con = engine).head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
ratings = ''' SELECT * 
                FROM ratings '''
pd.io.sql.read_sql(ratings, con = engine).head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
reviews = ''' SELECT * 
                FROM reviews '''
pd.io.sql.read_sql(reviews, con = engine).head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Итак, 5 таблиц, данные загружены. Можно приступать к заданиям.

## Задания

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [8]:
query = ''' 
            SELECT COUNT(book_id) as count 
            FROM books 
            WHERE publication_date > '2000-01-01'; 
'''
pd.io.sql.read_sql(query, con = engine)

count
0    819

**Вывод**: 819 книг вышло после 1-го января 2000 года.

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [9]:
query = '''
            SELECT 
                b.book_id, 
                b.title, 
                COUNT(DISTINCT rev.review_id) as reviews_count,  
                AVG(rat.rating) as avg_ratings        
            FROM 
                books AS b
            LEFT JOIN ratings AS rat ON rat.book_id = b.book_id 
            LEFT JOIN reviews AS rev ON rev.book_id = b.book_id 
            GROUP BY b.book_id;
'''
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  \
0          1                                       'Salem's Lot   
1          2                 1 000 Places to See Before You Die   
2          3  13 Little Blue Envelopes (Little Blue Envelope...   
3          4  1491: New Revelations of the Americas Before C...   
4          5                                               1776   
..       ...                                                ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997                        Xenocide (Ender's Saga  #3)   
997      998                                    Year of Wonders   
998      999                        You Suck (A Love Story  #2)   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...   

     reviews_count  avg_ratings  
0                2     3.666667  
1                1     2.500000  
2                3     4.666667  
3                2     4.500000  
4                4     4.000000  
..             ...          ...  
995              3     3.666667  
996              3     3.400000  
997              4     3.200000  
998              2     4.500000  
999              4     3.833333  

[1000 rows x 4 columns]

**Вывод:** Получилась таблица, с количеством обзоров (reviews_count) и средней оценки (avg_ratings) для каждой книги

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [10]:
query =  '''
            SELECT p.publisher,
            COUNT(b.book_id) as count            

            FROM books AS b
            LEFT JOIN publishers AS p ON p.publisher_id = b.publisher_id
            
            WHERE num_pages > 50
            GROUP BY p.publisher
            ORDER BY count DESC
            LIMIT 1;
'''
pd.io.sql.read_sql(query, con = engine)

publisher  count
0  Penguin Books     42

**Вывод**: **Pinguin Books** - издательство, выпустившее наибольшее количество книг толще 50 страниц (42 книги).

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [11]:
query = ''' 
            SELECT a.author, 
            AVG(rat.rating) as avg_rating

            FROM books AS b

            LEFT JOIN authors AS a on b.author_id = a.author_id
            LEFT JOIN ratings AS rat on b.book_id = rat.book_id

            WHERE b.book_id IN
            
                (SELECT b.book_id
                FROM books AS b
                LEFT JOIN ratings AS rat on b.book_id = rat.book_id
                GROUP BY b.book_id
                HAVING COUNT(rat.rating_id) >= 50)
                
            GROUP BY a.author
            ORDER BY avg_rating DESC
            LIMIT 1;
''' 

pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

**Вывод:** J.K. Rowling/Mary GrandPré - авторы с самой высокой средней оценкой книг (4,3).

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query = '''
            SELECT ROUND(SUM(r.count) / COUNT(r.username))  AS avg_count_reviews
            
            FROM
                (SELECT  rev.username,
                COUNT(rev.review_id) AS count
                FROM reviews AS rev
                WHERE rev.username IN
                    
                    (SELECT rat.username
                     FROM ratings AS rat
                     LEFT JOIN books AS b ON b.book_id = rat.book_id
                     GROUP BY  rat.username
                     HAVING COUNT(rat.rating_id) > 50)
                
            GROUP BY rev.username) AS r;

       '''
pd.io.sql.read_sql(query, con = engine)

avg_count_reviews
0               24.0

**Вывод**: 24 - среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Общий вывод

В ходе анализа базы данных для сервиса для чтения книг по подписке, выяснено:
- 819 книг вышло после 1-го января 2000 года;
- Получена таблица, с количеством обзоров и средней оценкой для каждой книги;
- Pinguin Books - издательство, выпустившее наибольшее количество книг толще 50 страниц (42 книги);
- J.K. Rowling/Mary GrandPré - авторы с самой высокой средней оценкой книг (4,3);
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.